include/mbedtls/bignum.h

```diff
@@ -124,18 +124,42 @@
         #endif /* !MBEDTLS_HAVE_INT64 */
         typedef  int64_t mbedtls_mpi_sint;
         typedef uint64_t mbedtls_mpi_uint;
+        #if !defined(MBEDTLS_MPI_UINT_BITS)
+            #define MBEDTLS_MPI_UINT_BITS 64
+        #endif /* MBEDTLS_MPI_UINT_BITS */
+        #if !defined(MBEDTLS_MPI_HALF_MASK)
+            #define MBEDTLS_MPI_HALF_MASK 0xffffffff
+        #endif /* MBEDTLS_MPI_HALF_MASK */
+        #if !defined(MBEDTLS_MPI_HALF_BITS)
+            #define MBEDTLS_MPI_HALF_BITS 32
+        #endif /* MBEDTLS_MPI_HALF_BITS */
+        #if !defined(MBEDTLS_MPI_UINT_MASK)
+            #define MBEDTLS_MPI_UINT_MASK 0x7fffffffffffffffLL
+        #endif /* MBEDTLS_MPI_UINT_MASK */
     #elif defined(__GNUC__) && (                         \
         defined(__amd64__) || defined(__x86_64__)     || \
         defined(__ppc64__) || defined(__powerpc64__)  || \
         defined(__ia64__)  || defined(__alpha__)      || \
         ( defined(__sparc__) && defined(__arch64__) ) || \
         defined(__s390x__) || defined(__mips64)       || \
         defined(__aarch64__) )
-        #if !defined(MBEDTLS_HAVE_INT64)
+    #if !defined(MBEDTLS_HAVE_INT64)
             #define MBEDTLS_HAVE_INT64
         #endif /* MBEDTLS_HAVE_INT64 */
         typedef  int64_t mbedtls_mpi_sint;
         typedef uint64_t mbedtls_mpi_uint;
+        #if !defined(MBEDTLS_MPI_UINT_BITS)
+            #define MBEDTLS_MPI_UINT_BITS 64
+        #endif /* MBEDTLS_MPI_UINT_BITS */
+        #if !defined(MBEDTLS_MPI_UINT_MASK)
```

Comment:
why is the order for MSC and GNUC not the same?

With so much in common maybe just set MBEDTLS_MPI_UINT_BITS to 64 in the compiler check and then process the rest of the defines based on this later?

Reply:
There are some platforms (i.e. Windows, where I first tested my changes) where the integer size is 32, not 64.

Reply:
I think what @sander-visser was suggesting is to set `MBEDTLS_MPI_UINT_BITS` to the appropriate value (64 or 32) here, but not the `UINT_MASK`, `HALF_BITS` and `HALF_MASK` macros. Then after this big "compiler/platform" is, have separate

```c
#if MBEDTLS_MPI_UINT_BITS == 64
#define MBEDTLS_MPI_HALF_MASK 0xffffffff
#define MBEDTLS_MPI_HALF_BITS 32
#define MBEDTLS_MPI_INT_MASK 0xffffffffffffffffLL
#else 
/* etc */
```
This would avoid repetition and make it easier to check that `_BITS` and `_MASK` macros are always consistent with each other.

Also, I'm not sure I understand the point of the `if !defined()` - if one of these macros is already defined, we have no idea if the defined value is consistent with what we're doing here, so if that happens we do want the compiler to emit an error about the redefinition.

Reply:
You're both right, it makes a lot more sense to write that the way you suggested.

https://github.com/ARMmbed/mbedtls/pull/3309/commits/f8a714ab729b72afb45609243dcf911529df5ec5 implements your feedback.